sdk/storage/Azure.Storage.Common/src/Shared/StorageBearerTokenChallengeAuthorizationPolicy.cs

```diff
@@ -81,6 +81,10 @@ private async ValueTask<bool> AuthorizeRequestOnChallengeInternalAsync(HttpMessa
                 // tenantId should be the guid as seen in this example: https://login.microsoftonline.com/72f988bf-86f1-41af-91ab-2d7cd011db47/oauth2/authorize
                 tenantId = new Uri(authUri).Segments[1].Trim('/');
 
+                string scope = AuthorizationChallengeParser.GetChallengeParameterFromResponse(message.Response, "Bearer", "resource_id");
+                scope += Constants.DefaultScope;
+                _scopes[0] = scope;
```

Comment:
This looks a little questionable to me.  The policy is expected to be thread-safe and shared across requests, but this is mutating instance state on a per-request basis.   (The `_scopes` variable is also annotated `readonly`, and even though that does not extend to the values of the array, it violates the principle of least surprise to see the elements change.)

While the `resource_id` is currently constant across all requests, I'm concerned about taking that dependency here.  

(As an aside, why is the `tenantId`  variable not named `_tenantId`?)

Reply:
See how ChallengeBasedAuthenticationPolicy does it in a threadsafe way here: 

https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/keyvault/Azure.Security.KeyVault.Shared/src/ChallengeBasedAuthenticationPolicy.cs#L20

Reply:
In running local live tests, I've noticed the current behavior is that the client will continue sending the invalid scope and getting a bearer challenge for each request it sends, and then sending a follow up request with the correct scope.

Could I just switch scope to a local variable, as @ryanerdmann suggested?

Reply:
It would only be referenced locally when a challenge is received. For non-challenge scenarios, you'll need to supply the new scope somehow. ChallengeBasedAuthenticationPolicy handles this in its override of `AuthorizeRequestAsync` which calls `AuthorizeRequestInternal`

Reply:
I have updated the PR to pass _scopes to the base.  We may still have a thread safety issue though.